### PR TITLE
add resource BMS instance

### LINKED ISSUE #1024: Feature Request: Add BMS resource support
Hi there,
I'd like to add BMS resource support on terraform side.

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

add resource BMS instance

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1024 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccBmsInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccBmsInstance_basic -timeout 360m -parallel 4
=== RUN   TestAccBmsInstance_basic
=== PAUSE TestAccBmsInstance_basic
=== CONT  TestAccBmsInstance_basic
--- PASS: TestAccBmsInstance_basic (824.09s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       824.154s
```

### CODE REVIEW

File: docs/resources/bms_instance.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+installed, the `admin_pass` field becomes invalid.
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the instance.
+    If omitted, the provider-level region will be used.
+    Changing this creates a new instance.
+
+* `name` - (Required, String) Specifies a unique name for the instance.
+    The name consists of 1 to 64 characters, including letters, digits, underscores (_), hyphens (-), and periods (.).
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: The length of name ranges from 1 to 63.

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+				Optional: true,
+				ForceNew: true,
+			},
+			"security_groups": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: The required parameters should precede the optional parameters.

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"op_svc_userid": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: Is it easier to understand using `user_id` naming?

File: docs/resources/bms_instance.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+* `flavor_id` - (Required, String, ForceNew) Specifies the flavor ID of the desired flavor for the instance.
+    Changing this creates a new instance.
+
+* `op_svc_userid` - (Required, String, ForceNew) Specifies the user ID. You can obtain the user ID
+    from My Credential on the management console. Changing this creates a new instance.
+
+* `user_data` - (Optional, String, ForceNew) Specifies the user data to be injected during the instance creation.
+    Text and text files can be injected.
+    Changing this creates a new server.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: make sure the naming rule (`server` or `instance`) in the document are the same (same problem elsewhere in the document).

File: docs/resources/bms_instance.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+* `admin_pass` - (Optional, String, ForceNew) Specifies the administrative password to assign to the instance.
+    Changing this creates a new instance.
+
+* `key_name` - (Optional, String, ForceNew) Specifies the name of a key pair to put on the instance.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: I think it is enough to name it as `keypair' and indicate in the doc that this parameter is the name of the keypair.

File: docs/resources/bms_instance.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    The key pair must already be created and associated with the tenant's account.
+    Changing this creates a new instance.
+
+* `security_groups` - (Optional, List, ForceNew) Specifies a array of one or more security group ids to associate
```
Comment (Lance52259) [inline docs/resources/bms_instance.md:122]: The description of `forceNew` is missing.

File: docs/resources/bms_instance.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    with the instance.
+
+* `availability_zone` - (Required, String, ForceNew) Specifies the availability zone in which to create
+    the instance. Please following [reference](https://developer.huaweicloud.com/intl/en-us/endpoint?BMS)
+    for the values. Changing this creates a new instance.
+
+* `vpc_id` - (Required, List, ForceNew) Specifies id of vpc in which to create the instance.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: This type is wrong and missing `forceNew` description.

File: docs/resources/bms_instance.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - A resource ID in UUID format.
+* `host_id` - The host ID of the instance.
+* `status` - The status of the instance.
+* `description` - The description of the instance.
+* `image_name` - The image_name of the instance.
+* `public_ip` - The EIP address that is associted to the instance.
+
```
Comment (Lance52259) [inline docs/resources/bms_instance.md:229]: `nics/mac_address` and `nics/port_id` are missing. please check if there are other missing parameters.

File: docs/resources/bms_instance.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    for the values. Changing this creates a new instance.
+
+* `vpc_id` - (Required, List, ForceNew) Specifies id of vpc in which to create the instance.
+
+* `nics` - (Required, List, ForceNew) Specifies an array of one or more networks to attach to the instance.
+    The network object structure is documented below.
+    Changing this creates a new instance.
+
+* `eip_id` - (Optional, String, ForceNew) The ID of the EIP. Changing this parameter will create a new resource.
+
```
Comment (Lance52259) [inline docs/resources/bms_instance.md:126]: extra blank lines (same problem elsewhere in the document).

File: docs/resources/bms_instance.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+-> **Note:** If the eip_id parameter is configured, you do not need to configure the bandwidth parameters:
+  `iptype`, `eip_charge_mode`, `bandwidth_size`, `share_type` and `bandwidth_charge_mode`.
+
+* `iptype` - (Optional, String, ForceNew) Elastic IP type. Changing this parameter will create a new resource.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: is `ip_type`?

File: docs/resources/bms_instance.md [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+* `eip_charge_mode` - (Optional, String, ForceNew) Elastic IP billing type. Changing this parameter will create a new resource.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: The suffix name should be `charging_mode`.

File: docs/resources/bms_instance.md (hunk 10 above)
Comment (Lance52259) [reply docs/resources/bms_instance.md:132]: The descrption of `forceNew` need to update

File: docs/resources/bms_instance.md [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+* `sharetype` - (Optional, String, ForceNew) Bandwidth sharing type. Changing this parameter will create a new resource.
+
+* `bandwidth_size` - (Optional, Int, ForceNew) Bandwidth size. Changing this parameter will create a new resource.
+
+* `bandwidth_charge_mode` - (Optional, String, ForceNew) Bandwidth billing type. Changing this parameter will create a new resource.
+
+
```
Comment (Lance52259) [inline docs/resources/bms_instance.md:142]: dito

File: docs/resources/bms_instance.md [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+* `system_disk_type` - (Optional, String, ForceNew) Specifies the system disk type of the instance.
+  For details about disk types, see
+  [Disk Types and Disk Performance](https://support.huaweicloud.com/en-us/productdesc-evs/en-us_topic_0014580744.html)
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: The link should be international link

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"eip_id"},
+			},
+			"system_disk_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"SATA", "SAS", "SSD", "GPSSD",
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: Is the value of `SATA` valid?

File: docs/resources/bms_instance.md [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    Available options are:
+	* `SSD`: ultra-high I/O disk type.
+	* `GPSSD`: general purpose SSD disk type.
+	* `SAS`: high I/O disk type.
+    Changing this creates a new instance.
+
+* `system_disk_size` - (Optional, int, ForceNew) Specifies the system disk size in GB, The value range is 1 to 1024.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: The minimum of the system disk size should be 40.

File: docs/resources/bms_instance.md [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+    Shrinking the disk is not supported.
+
+* `data_disks` - (Optional, String, ForceNew) Specifies an array of one or more data disks to attach to the instance.
```
Comment (Lance52259) [inline docs/resources/bms_instance.md]: The type of `data_disks` should be `List`.
You can consider whether the maximum number of `data_disks` needs to be declared.

File: huaweicloud/resource_huaweicloud_bms_instance_test.go [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+		return nil
+	}
+}
+
+func testAccBmsInstance_basic(rName string) string {
+	return fmt.Sprintf(`
+
+resource "huaweicloud_compute_keypair" "test" {
+  name = "%s"
+  public_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDAjpC1hwiOCCmKEWxJ4qzTTsJbKzndLo1BCz5PcwtUnflmU+gHJtWMZKpuEGVi29h0A/+ydKek1O18k10Ff+4tyFjiHDQAT9+OfgWf7+b1yK+qDip3X1C0UPMbwHlTfSGWLGZquwhvEFx9k3h/M+VtMvwR1lJ9LUyTAImnNjWG7TAIPmui30HvM2UiFEmqkr4ijq45MyX2+fLIePLRIFuu1p4whjHAQYufqyno3BS48icQb4p6iVEZPo4AE2o9oIyQvj2mx4dk5Y8CgSETOZTYDOR3rU2fZTRDRgPJDH9FWvQjF5tA0p3d9CoWWd2s6GKKbfoUIi8R/Db1BSPJwkqB jrp-hp-pc"
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance_test.go]: `public_key` include hard code and it's used to import the keypair, can we use lifecycle to ignore the state check?

File: huaweicloud/resource_huaweicloud_bms_instance_test.go [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    name        = "%s"
+    size        = 8
+    share_type  = "PER"
+    charge_mode = "traffic"
+  }
+}
+
+resource "huaweicloud_bms_instance" "test" {
+  name                  = "%s"
+  image_id              = "519ea918-1fea-4ebc-911a-593739b1a3bc"
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance_test.go:136]: can we get this image by `huaweicloud_images_image` resource?

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 19]
```diff
@@ -0,0 +1,595 @@
...
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"nics": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				MinItems: 1,
+				MaxItems: 12,
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: Is this maxItems value right?

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 20]
```diff
@@ -0,0 +1,595 @@
...
+			ChargingMode:        d.Get("charging_mode").(string),
+			PeriodType:          d.Get("period_unit").(string),
+			PeriodNum:           d.Get("period").(int),
+			IsAutoPay:           "true",
+			IsAutoRenew:         d.Get("auto_renew").(string),
+			EnterpriseProjectId: GetEnterpriseProjectID(d, config),
+		},
+	}
+
+	if hasFilledOpt(d, "eip_id") || hasFilledOpt(d, "iptype") {
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go:322]: hasFilledOpt() is not suggest used

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 21]
```diff
@@ -0,0 +1,595 @@
...
+		createOpts.RootVolume = &volRequest
+	}
+
+	if hasFilledOpt(d, "tags") {
+		tagRaw := d.Get("tags").(map[string]interface{})
+		taglist := utils.ExpandResourceTags(tagRaw)
+		createOpts.ServerTags = taglist
+	}
+
+	log.Printf("[DEBUG] Create Options: %#v", createOpts)
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: This log will contain sensitive information (admin_pass)

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 22]
```diff
@@ -0,0 +1,595 @@
...
+
+	n, err := baremetalservers.CreatePrePaid(bmsClient, createOpts).ExtractOrderResponse()
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud baremetalserver: %s", err)
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: suggest to split the `baremetalserver`

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 23]
```diff
@@ -0,0 +1,595 @@
...
+		return err
+	}
+	server_id := entity.(string)
+
+	if server_id != "" {
+		d.SetId(server_id)
+		return resourceBmsInstanceRead(d, meta)
+	}
+
+	return fmt.Errorf("Unexpected conversion error in resourceBmsInstanceCreate.")
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: Error message should not contains period.

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 24]
```diff
@@ -0,0 +1,595 @@
...
+	for i, raw := range rawSecGroups {
+		secgroups[i] = baremetalservers.SecurityGroup{
+			ID: raw.(string),
+		}
+	}
+	return secgroups
+}
+
+func flattenBmsInstanceNicsV1(
+	d *schema.ResourceData, meta interface{}, addresses map[string][]baremetalservers.Address) []map[string]interface{} {
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: The position of the line break is generally not before the first parameter.

File: docs/resources/bms_instance.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 2 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: docs/resources/bms_instance.md (hunk 4 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: docs/resources/bms_instance.md (hunk 5 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: docs/resources/bms_instance.md (hunk 6 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md:122]: fixed

File: docs/resources/bms_instance.md (hunk 7 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: docs/resources/bms_instance.md (hunk 8 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md:229]: fixed

File: docs/resources/bms_instance.md (hunk 9 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md:126]: fixed

File: docs/resources/bms_instance.md (hunk 10 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: `iptype` is also used in resource cce node, it's better to be consistent

File: docs/resources/bms_instance.md (hunk 11 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: `charge_mode` is also used in resource cce node, it's better to be consistent

File: docs/resources/bms_instance.md (hunk 13 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 14 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: docs/resources/bms_instance.md (hunk 15 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: You can check the console. Accturaly, the value of size ranges from 0 to 1024. But I think 0 is useless.

File: docs/resources/bms_instance.md (hunk 16 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance_test.go (hunk 18 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance_test.go:136]: unless use exact image name, there is no better way for now

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 19 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 24 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 23 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 22 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 21 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance_test.go (hunk 17 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance_test.go]: fixed

File: docs/resources/bms_instance.md (hunk 15 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: The system disk size must be greater than or equal to the minimum system disk size of the image. I added this description.

File: docs/resources/bms_instance.md [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  tags = {
+    foo = "bar"
+    key = "value"
+  }
+}
+```
+
+`user_data` can come from a variety of sources: inline, read in from the `file`
```
Comment (ShiChangkuo) [inline docs/resources/bms_instance.md]: put this line after L114

File: docs/resources/bms_instance.md [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+function, or the `template_cloudinit_config` resource.
+
+## Argument Reference
+
+-> **NOTE:** If the `user_data` field is specified for a Linux ECS that is created using an image with Cloud-Init
```
Comment (ShiChangkuo) [inline docs/resources/bms_instance.md]: put this line after L119

File: docs/resources/bms_instance.md [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+* `name` - (Required, String) Specifies a unique name for the instance.
+    The name consists of 1 to 63 characters, including letters, digits, underscores (_), hyphens (-), and periods (.).
+
+* `image_id` - (Required, String, ForceNew) Specifies the image ID of the desired image for the instance.
+    Changing this creates a new instance.
+
+* `flavor_id` - (Required, String, ForceNew) Specifies the flavor ID of the desired flavor for the instance.
+    Changing this creates a new instance.
+
+* `op_svc_userid` - (Required, String, ForceNew) Specifies the user ID. You can obtain the user ID
```
Comment (ShiChangkuo) [inline docs/resources/bms_instance.md]: suggest to rename this parameter to `user_id`, and we can get the value from `HW_USER_ID`

File: huaweicloud/provider_test.go [hunk 28]
```diff
@@ -151,6 +152,12 @@ func testAccAsConfigPreCheck(t *testing.T) {
 	}
 }
 
+func testAccPreCheckBms(t *testing.T) {
+	if HW_OP_SVC_USERID == "" {
```
Comment (ShiChangkuo) [inline huaweicloud/provider_test.go]: please use `HW_USER_ID` instead

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 29]
```diff
@@ -0,0 +1,589 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"flavor_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"op_svc_userid": {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: rename to `user_id`

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 30]
```diff
@@ -0,0 +1,589 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"nics": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				MinItems: 1,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: MinItems is unnecessary for required parameters

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 31]
```diff
@@ -0,0 +1,589 @@
...
+						return hex.EncodeToString(hash[:])
+					default:
+						return ""
+					}
+				},
+			},
+			"admin_pass": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: mark it as sensitive.

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 32]
```diff
@@ -0,0 +1,589 @@
...
+			},
+			"key_pair": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go:125]: admin_pass and key_pair should be alternative?

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 33]
```diff
@@ -0,0 +1,589 @@
...
+				ConflictsWith: []string{
+					"iptype", "eip_charge_mode", "bandwidth_charge_mode", "bandwidth_size", "sharetype",
+				},
+			},
+			"iptype": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"eip_id"},
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go:152]: shall we add `ValidateFunc` attribute?

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 34]
```diff
@@ -0,0 +1,589 @@
...
+				RequiredWith: []string{
+					"eip_charge_mode", "sharetype", "bandwidth_size",
+				},
+			},
+			"eip_charge_mode": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"eip_id"},
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go:164]: ditto

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 35]
```diff
@@ -0,0 +1,589 @@
...
+				RequiredWith: []string{
+					"iptype", "sharetype", "bandwidth_size",
+				},
+			},
+			"sharetype": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: ditto

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 36]
```diff
@@ -0,0 +1,589 @@
...
+				ConflictsWith: []string{"eip_id"},
+				RequiredWith: []string{
+					"iptype", "eip_charge_mode", "sharetype",
+				},
+			},
+			"bandwidth_charge_mode": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: ditto

File: docs/resources/bms_instance.md [hunk 37]
```diff
@@ -0,0 +1,218 @@
...
+    The data_disks object structure is documented below. A maximum of 59 disks can be mounted.
+    Changing this creates a new instance.
+
+* `tags` - (Optional, Map, ForceNew) Specifies the key/value pairs to associate with the instance.
+    Changing this creates a new instance.
+
+* `enterprise_project_id` - (Optional, String, ForceNew) Specifies a unique id in UUID format of enterprise project .
+    Changing this creates a new instance.
+
+* `charging_mode` - (Optional, String, ForceNew) Specifies the charging mode of the instance.
```
Comment (ShiChangkuo) [inline docs/resources/bms_instance.md:181]: we should update the docs if only *prePaid* is supported.

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 38]
```diff
@@ -0,0 +1,589 @@
...
+	config := meta.(*config.Config)
+	bmsClient, err := config.BmsV1Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud compute client: %s", err)
+	}
+	serverID := d.Id()
+	publicIP := d.Get("public_ip").(string)
+
+	resourceIDs := make([]string, 0, 2)
+	if serverID != "" {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go]: serverID will never be empty before deleting it

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 39]
```diff
@@ -0,0 +1,589 @@
...
+			}
+			nics = append(nics, v)
+		}
+	}
+
+	log.Printf("[DEBUG] flattenInstanceNicsV1: %#v", nics)
+	return nics
+}
+
+func bmsPublicIP(server *baremetalservers.CloudServer) string {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go:594]: can we use the similar function in `huaweicloud_compute_instance`?

File: huaweicloud/resource_huaweicloud_bms_instance.go [hunk 40]
```diff
@@ -0,0 +1,589 @@
...
+	secgroups := make([]baremetalservers.SecurityGroup, len(rawSecGroups))
+	for i, raw := range rawSecGroups {
+		secgroups[i] = baremetalservers.SecurityGroup{
+			ID: raw.(string),
+		}
+	}
+	return secgroups
+}
+
+func flattenBmsInstanceNicsV1(d *schema.ResourceData, meta interface{},
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_bms_instance.go:553]: can we use the similar function in `huaweicloud_compute_instance`?

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]

File: docs/resources/bms_instance.md (hunk 25 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: docs/resources/bms_instance.md (hunk 26 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: docs/resources/bms_instance.md (hunk 27 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 29 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/provider_test.go (hunk 28 above)
Comment (Jason-Zhang9309) [reply huaweicloud/provider_test.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 30 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 31 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 32 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:125]: fixed

File: docs/resources/bms_instance.md (hunk 37 above)
Comment (Jason-Zhang9309) [reply docs/resources/bms_instance.md:181]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 38 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 39 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:562]: fixed

Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:594]: the type of param `server` is different

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 40 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:553]: the type of param `address` is different

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 33 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:152]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 34 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go:164]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 35 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_bms_instance.go (hunk 36 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_bms_instance.go]: fixed

**Pull request merged**